packages/core/src/makeInputBinding.ts

```diff
@@ -1,81 +1,75 @@
 import { Shiny } from "./utils";
-import { Constructor } from "./utils";
 
 /**
- * A custom element that extends this interface will be treated as an input
- * binding by Shiny when paired with `makeInputBinding()`.
+ * Make a custom input binding
+ *
+ * @param name The name of the binding. This will be used to register the binding with
+ * Shiny. Must be unique.
+ * @param selector The selector to use to find the element to bind to. Defaults to looking for
+ * a class with the same name as the binding.
+ * @param setup A function that will be called when the element is first bound. Arguments are
+ * the element to bind to and a function that should be called by the input binding when there
+ * is a new value.
+ *
  */
-export interface CustomElementInput<T = string> extends HTMLElement {
-  id: string;
-  value: T;
-  /**
-   * Function to let Shiny know it should look for a new value.
-   * @param allowDeferred Should the value be immediately updated wait to the next event loop?
-   */
-  notifyBindingOfChange: (allowDeferred?: boolean) => void;
-}
-
-/**
- * Given a tag name for a custom element that is a CustomElementInput<T>, this
- * will hook up the proper input binding and register it with Shiny.
- * @param tagName Name of the tag that corresponds to the input binding
- * @param el The custom element that extends `CustomElementInput<T>`.
- * @param opts Options for the output binding
- * @param opts.registerElement Whether to register the webcomponent used for the
- * output. Defaults to true.
- * @param opts.type The type of the input. This is used for type-checking the
- * input within Shiny. Defaults to null.
- */
-export function makeInputBinding<
-  T,
-  El extends CustomElementInput<T> = CustomElementInput<T>
->(
-  tagName: string,
-  el: Constructor<El>,
-  opts: {
-    registerElement?: boolean;
-    type?: string | null;
-  } = { registerElement: true, type: null }
-) {
+export function makeInputBinding<T>({
+  name,
+  selector = `.${name}`,
+  setup,
+}: {
+  name: string;
+  selector?: string;
+  setup: (
+    el: HTMLElement,
+    onNewValue: (x: T, allowDeferred?: boolean) => void
```

Comment:
Change here and in examples

```suggestion
    updateValue: (x: T, allowDeferred?: boolean) => void
```